File: shunting_yard.pl
#!/usr/bin/perl

use strict;
use warnings;

use Data::Dumper;
use Carp;

my $operators = {
  '+' => { 'precedence' => 2, 'associativity' => 'left', },
  '-' => { 'precedence' => 2, 'associativity' => 'left', },
  '*' => { 'precedence' => 3, 'associativity' => 'left', },
  '/' => { 'precedence' => 3, 'associativity' => 'left', },
  '^' => { 'precedence' => 4, 'associativity' => 'right', },
  };

sub shunting_yard
{
    my ($expression, $opts) = @_;

    $opts = {} unless $opts;

    my $in = [];
    my $out = [];
    my $op_stack = [];

    # split string into array (tokenize)
    my $in_str = [split("", $expression)];
    my $token = "";

    my $previous_token = "";

    CHAR: while (my $char = shift(@{$in_str})) {
        next CHAR if $char =~ /\s/;
        
        if ($operators->{$previous_token} && ($char eq "+" || $char eq "-")) {
            # previos was [+-*/^], current is + or -
            $token = $token . $char;
            $previous_token = "";
        } elsif ($operators->{$char} || $char eq "(" || $char eq ")") {
            if ($token) {
                push (@{$in}, $token);
            }
            $token = "";

            push (@{$in}, $char);

            $previous_token = $char;
        } else {
            $token = $token . $char;
            $previous_token = "";
        }
    }
    if ($token) {
        push (@{$in}, $token);
    }

    my $infix = [@{$in}];

    # convert array from infix to rpn
    TOKEN: while (my $token = shift(@{$in}) ) {
        if ($operators->{$token}) {
            CHECK_OP_STACK: while (scalar(@{$op_stack})) {
                my $top_op = $op_stack->[scalar(@{$op_stack}) - 1];
    
                if ($operators->{$top_op} &&
                    (
                        $operators->{$token}->{'precedence'} == $operators->{$top_op}->{'precedence'} &&
                        $operators->{$token}->{'associativity'} eq 'left' ||
                        $operators->{$token}->{'precedence'} < $operators->{$top_op}->{'precedence'}
                    )
                    ) {
                    
                    push (@{$out}, pop(@{$op_stack}));
                } else {
                    last CHECK_OP_STACK;
                }
            }

            push (@{$op_stack}, $token);
        } elsif ($token eq "(") {
            push (@{$op_stack}, $token);
        } elsif ($token eq ")") {
            my $found_matching_paren;
            CHECK_OP_STACK: while (scalar(@{$op_stack})) {
                my $top_op = $op_stack->[scalar(@{$op_stack}) - 1];
                
                if ($top_op ne "(") {
                    push (@{$out}, pop(@{$op_stack}));
                } else {
                    $found_matching_paren = 1;
                    pop(@{$op_stack});
                    last CHECK_OP_STACK;
                }
            }
            Carp::confess("Invalid expression: can't find matching parenthesis")
                if !$found_matching_paren;
        } else {
            push @{$out}, $token;
        }
    }

    while (scalar(@{$op_stack})) {
        push (@{$out}, pop(@{$op_stack}));
    }

    if ($opts->{'debug'}) {
        print "ORIGINAL: " . $expression . "\n";
        print "TOKENIZED: " . join(" ", @{$infix}) . "\n";
        print "RPN: " . join(" ", @{$out}) . "\n";
    }

    return {
        'original' => $expression,
        'infix' => $infix,
        'rpn' => $out,
        };
}

sub calc_rpn
{
    my ($rpn, $opts) = @_;

    $opts = {} unless $opts;

    my $calc = [];

    while (my $token = shift(@{$rpn})) {
        if ($operators->{$token}) {
            my $tmp_value_1 = pop (@{$calc});
            my $tmp_value_2 = pop (@{$calc});

            if ($token eq "+") {
                push (@{$calc}, $tmp_value_2 + $tmp_value_1);
            } elsif ($token eq "-") {
                push (@{$calc}, $tmp_value_2 - $tmp_value_1);
            } elsif ($token eq "*") {
                push (@{$calc}, $tmp_value_2 * $tmp_value_1);
            } elsif ($token eq "/") {
                push (@{$calc}, $tmp_value_2 / $tmp_value_1);
            } elsif ($token eq "^") {
                push (@{$calc}, $tmp_value_2 ** $tmp_value_1);
            } else {
                Carp::confess("Programmer error: operator [$token] is not implemented.");
            }
        } else {
            push (@{$calc}, $token);
        }
    }

    if ($opts->{'debug'}) {
        print $calc->[0] . "\n";
    }

    return $calc->[0];
}

my $r;
my $r1;

if (!$ARGV[0]) {
    printf("usage: shuntint_yard.pl <test|EXPRESION>\n");
    exit(0);
}

if ($ARGV[0] eq 'test') {
    my $tests = [
        { 'expression' => "2+3",             'expected_rpn' => "2 3 +",                     'expected_result' => 5, },
        { 'expression' => "2+3*5-12",        'expected_rpn' => "2 3 5 * + 12 -",            'expected_result' => 5, },
        { 'expression' => "3+4*2/(1-5)",     'expected_rpn' => "3 4 2 * 1 5 - / +",         'expected_result' => 1, },
        { 'expression' => "3+4*2/(1-5)^2^3", 'expected_rpn' => "3 4 2 * 1 5 - 2 3 ^ ^ / +", 'expected_result' => (3+1/8192), },
        { 'expression' => "1+-1",            'expected_rpn' => "1 -1 +",                    'expected_result' => 0, },
        { 'expression' => "3+2*5-2/2",       'expected_rpn' => "3 2 5 * + 2 2 / -",         'expected_result' => 12, },
        { 'expression' => "2^3+3^2",         'expected_rpn' => "2 3 ^ 3 2 ^ +",             'expected_result' => 17, },
        { 'expression' => "2^(3+1)-(1+1)*-1",'expected_rpn' => "2 3 1 + ^ 1 1 + -1 * -",    'expected_result' => 18, },
        { 'expression' => "(((1)))",         'expected_rpn' => "1",                         'expected_result' => 1, },
        ];

    my $i = 0;
    foreach my $test (@{$tests}) {
        $i++;

        $r = shunting_yard($test->{'expression'}, {'debug' => 1});
        if (join(" ", @{$r->{'rpn'}}) eq $test->{'expected_rpn'}) {
            print "test $i.1 passed\n";
        } else {
            Carp::confess("test $i.1 failed; expected rpn [" . $test->{'expected_rpn'} . "], got: " . Data::Dumper::Dumper($r));
        }
    
        $r1 = calc_rpn($r->{'rpn'}, {'debug' => 1});
        if ($r1 == $test->{'expected_result'}) {
            print "test $i.2 passed\n";
        } else {
            Carp::confess("test $i.2 failed; expected result [" . $test->{'expected_result'} . "], got " . Data::Dumper::Dumper($r1));
        }
        printf("\n");
    }
}
else {
    $r = shunting_yard($ARGV[0], {'debug' => 1});
    $r1 = calc_rpn($r->{'rpn'}, {'debug' => 1});
}

File: shunting_yard.c
//
// shunting yard implementation
//

#include <mystd.h>
#include <stack.h>

// isspace
#include <ctype.h>

// pow
#include <math.h>

#define LEFT 1
#define RIGHT 2

bool is_operator (char ch)
{
    if (ch == '+' ||
        ch == '-' ||
        ch == '*' ||
        ch == '/' ||
        ch == '^'
        ) {
        return true;
    } else {
        return false;
    }
}

bool is_operator_str (char *token)
{
    if (strlen(token) == 1 && is_operator(token[0])) {
        return true;
    } else {
        return false;
    }
}

int op_precedence (char ch)
{
    if (ch == '+' || ch == '-') {
        return 2;
    } else if (ch == '*' || ch == '/') {
        return 3;
    } else if (ch == '^') {
        return 4;
    } else {
        die("Programmer error: op_precedence got invalid operator [%c]", ch);
    }
  
    return 0;
}

int op_associativity (char ch)
{
    if (ch == '+' || ch == '-') {
        return LEFT;
    } else if (ch == '*' || ch == '/') {
        return LEFT;
    } else if (ch == '^') {
        return RIGHT;
    } else {
        die("Programmer error: op_associativity got invalid operator [%c]", ch);
    }

    return 0;
}

void shunting_yard(char *expression, stack *out)
{
    int expression_length = strlen(expression);
    if (expression_length < 1) {
        die("Programmer error: shunting_yard expects non-empty expression");
    }

    stack in_tokenized;
    stack op_stack;
    stack_create(&in_tokenized, "input tokenized", out->max_size);
    stack_create(&op_stack, "operator stack", out->max_size);

    char previous_token[STRUCT_NAME_LENGTH] = "";
    char token[STRUCT_NAME_LENGTH] = "";
    
    char current_char_str[2];
    char current_char;
    int i=0;

    // split string into array (tokenize) -- actually I use stack instead of array,
    // which implies reversing it before I can access elements sequentially 
    CHAR: while(i < expression_length) {
        current_char = expression[i];

        strncpy(current_char_str, &expression[i], 1); 
        current_char_str[1] = 0;      

        if (isspace(current_char)) {
            i++;
            goto CHAR;
        }

        if (strlen(previous_token) > 0 && (current_char == '+' || current_char == '-')) {
            strncat(token, &current_char, 1);
            previous_token[0] = 0;
        } else if (
            is_operator(current_char) == true ||
            current_char == '(' ||
            current_char == ')'
            ) {
            if (strlen(token) > 0) {
                stack_push(&in_tokenized, token);
            }
            token[0] = 0;

            stack_push(&in_tokenized, &current_char_str[0]);
            strcpy(previous_token, current_char_str);
        } else {
            strncat(token, &current_char, 1);
            previous_token[0] = 0;
        }

        i++;
    }
    if (strlen(token) > 0) {
        stack_push(&in_tokenized, token);
    }

    stack in_tokenized_reversed;
    stack_create(&in_tokenized_reversed, "input tokenized, reversed", in_tokenized.max_size);
    char *tmp_token;

    // reverse input stack, convert array from infix to rpn
    while (stack_is_empty(&in_tokenized) == false) {
        tmp_token = stack_pop(&in_tokenized);
        stack_push(&in_tokenized_reversed, tmp_token);
    }
    stack_free(&in_tokenized);

    TOKEN: while (stack_is_empty(&in_tokenized_reversed) == false) {
        tmp_token = stack_pop(&in_tokenized_reversed);
//        printf("%s\n", tmp_token);

        if (is_operator_str(tmp_token)) {
            bool break_CHECK_OP_STACK_1 = false;
            CHECK_OP_STACK_1: while (stack_is_empty(&op_stack) == false && break_CHECK_OP_STACK_1 == false) {
                char *tmp_top_op = stack_top(&op_stack);

                if (is_operator_str(tmp_top_op) && 
                    (
                        (op_precedence(tmp_token[0]) == op_precedence(tmp_top_op[0]) && op_associativity(tmp_token[0]) == LEFT) ||
                        op_precedence(tmp_token[0]) < op_precedence(tmp_top_op[0])
                    )
                    ) {
                    
                    stack_push(out, stack_pop(&op_stack));
                } else {
                    break_CHECK_OP_STACK_1 = true;
                    goto CHECK_OP_STACK_1;
                }
            }

            stack_push(&op_stack, tmp_token);
        } else if (tmp_token[0] == '(') {
            stack_push(&op_stack, tmp_token);
        } else if (tmp_token[0] == ')') {
            bool found_matching_paren = false;
            bool break_CHECK_OP_STACK_2 = false;
            CHECK_OP_STACK_2: while (stack_is_empty(&op_stack) == false && break_CHECK_OP_STACK_2 == false) {
                char *tmp_top_op = stack_top(&op_stack);

                if (tmp_top_op[0] != '(') {
                    stack_push(out, stack_pop(&op_stack));
                } else {
                    found_matching_paren = true;
                    stack_pop(&op_stack);
                    break_CHECK_OP_STACK_2 = true;
                    goto CHECK_OP_STACK_2;
                }
            }
            if (found_matching_paren == false) {
                die("Programmer error: shunting_yard got invalid expression, unbalanced parenthesis: [%s]", expression);
            }
        } else {
            stack_push(out, tmp_token);
        }

        goto TOKEN;
    }
    while (stack_is_empty(&op_stack) == false) {
        stack_push(out, stack_pop(&op_stack));
    }

    stack_free(&op_stack);
    stack_free(&in_tokenized_reversed);
}

float calc_rpn(stack *rpn)
{
    stack rpn_reversed;
    stack calc;

    stack_create(&rpn_reversed, "rpn reversed", rpn->max_size);
    stack_create(&calc, "calculator", rpn->max_size);

    while(stack_is_empty(rpn) == false) {
        stack_push(&rpn_reversed, stack_pop(rpn));
    }

    char *token;
    while (stack_is_empty(&rpn_reversed) == false) {
        token = stack_pop(&rpn_reversed);
        stack_push(rpn, token);

        if (is_operator_str(token)) {
            char *op1 = stack_pop(&calc);
            char *op2 = stack_pop(&calc);
        
            errno = 0;
            float op1_f = strtod(op1, (char **) NULL);
            if (errno != 0) {
                die_explaining_errno("not a number [%s] while calculating rpn", op1);
            }
            
            errno = 0;
            float op2_f = strtod(op2, (char **) NULL);
            if (errno != 0) {
                die_explaining_errno("not a number [%s] while calculating rpn", op2);
            }

            if (strcmp(token, "+") == 0) {
                token = (char *) malloc(MAX_MSG_SIZE);
                sprintf(token, "%f", op2_f + op1_f);
            } else if (strcmp(token, "-") == 0) {
                token = (char *) malloc(MAX_MSG_SIZE);
                sprintf(token, "%f", op2_f - op1_f);
            } else if (strcmp(token, "*") == 0) {
                token = (char *) malloc(MAX_MSG_SIZE);
                sprintf(token, "%f", op2_f * op1_f);
            } else if (strcmp(token, "/") == 0) {
                token = (char *) malloc(MAX_MSG_SIZE);
                sprintf(token, "%f", op2_f / op1_f);
            } else if (strcmp(token, "^") == 0) {
                token = (char *) malloc(MAX_MSG_SIZE);
                sprintf(token, "%f", pow(op2_f, op1_f));
            } else {
                die("Programmer error: operator [%s] is not implemented", token);
            }

            stack_push(&calc, token);
            free(token);
        } else {
            stack_push(&calc, token);
        }
    }

    token = stack_pop(&calc);
    errno = 0;
    float result = strtod(token, (char **) NULL);
    if (errno != 0) {
        die_explaining_errno("not a number [%s] while evaluating result", token);
    }
    
    stack_free(&rpn_reversed);
    stack_free(&calc);
    
    return result;
}

bool shunting_yard_test()
{
    struct struct_test {
        char *expression;
        float value;
    } tests[] = {
        { "2+3", 5 },
        { "2+3*5-12", 5 },
        { "3+4*2/(1-5)", 1 },
        { "3+4*2/(1-5)^2^3", 3+1.0/8192 },
        { "1+-1", 0 },
        { "3+2*5-2/2", 12 },
    };

    bool failed = false;
    stack rpn;
    float f;

    int n_tests = sizeof(tests) / sizeof(struct struct_test);

    int i;
    for (i=0; i<n_tests; i++) {
        stack_create(&rpn, "RPN", 100);
        shunting_yard(tests[i].expression, &rpn);
        f = calc_rpn(&rpn);

        if (f == tests[i].value) {
            printf("test %d passed: %s == %f\n", i+1, tests[i].expression, tests[i].value);
        } else {
            printf("test %d failed: %s: got %f, expected %f\n", i+1, tests[i].expression, f, tests[i].value);
            failed = true;
        }
        
        stack_free(&rpn);
    }

    if (failed == true) {
        return false;
    } else {
      return true;
    }
}

int main(int argc, char *argv[]) {
    if (argc == 2 && strcmp("test", argv[1]) == 0) {
        if (shunting_yard_test() == true) {
            printf("all tests are ok.\n");
        }

        return 0;
    } else {
        char *expression;
        
        int arg_size = 0;
        int i;
        for (i=1; i<argc; i++) {
            arg_size = arg_size + strlen(argv[i]);
        }
        
        expression = (char *) malloc(arg_size + 1);
        for (i=1; i<argc; i++) {
            strcat(expression, argv[i]);
        }

        stack rpn;
        stack_create(&rpn, "RPN", 100);
        shunting_yard(expression, &rpn);
        float f = calc_rpn(&rpn);
        stack_dump(&rpn);

        printf("expression: %s\n", expression);
        printf("value: %f\n", f);
    }

    return 0;
}
